advisor/src/main/kotlin/advisors/OssIndex.kt

```diff
@@ -0,0 +1,173 @@
+                    }
+                }
+
+                packageUrl
+            }
+
+            try {
+                val componentDetails = mutableMapOf<String, OssIndexService.Component>()
+
+                components.chunked(REQUEST_CHUNK_SIZE).forEach { component ->
+                    val serviceCall = withContext(advisorDispatcher) {
+                        service.getComponentReport(OssIndexService.ComponentsRequest(component))
+                    }
+
+                    val requestResults = execute(serviceCall).associateBy {
+                        it.coordinates
+                    }
+
+                    componentDetails += requestResults
+                }
+
+                val endTime = Instant.now()
+
+                packages.mapNotNullTo(mutableListOf()) { pkg ->
+                    componentDetails[pkg.id.toPurl()]?.takeUnless {
+                        it.vulnerabilities.isEmpty()
+                    }?.let { details ->
+                        pkg to listOf(
+                            AdvisorResult(
+                                details.vulnerabilities.map { it.toVulnerability() },
+                                AdvisorDetails(advisorName),
+                                AdvisorSummary(startTime, endTime)
+                            )
+                        )
+                    }
+                }.toMap()
+
+            } catch (e: IOException) {
+                e.showStackTrace()
+
+                val now = Instant.now()
+                packages.associateWith {
+                    listOf(
+                        AdvisorResult(
+                            vulnerabilities = emptyList(),
+                            advisor = AdvisorDetails(advisorName),
+                            summary = AdvisorSummary(
+                                startTime = now,
+                                endTime = now,
+                                issues = listOf(
+                                    createAndLogIssue(
+                                        source = advisorName,
+                                        message = "Failed to retrieve security vulnerabilities from $advisorName: " +
+                                                e.collectMessagesAsString()
+                                    )
+                                )
+                            )
+                        )
+                    )
+                }
+            }
+        }
+    }
+
+    private fun OssIndexService.Vulnerability.toVulnerability(): Vulnerability {
+        val browseUrl = URL(reference)
+
+        return Vulnerability(displayName, cvssScore, browseUrl)
+    }
+
+    /**
+     * Execute an HTTP request specified by the given [call]. The response status is checked. If everything went
+     * well, the marshalled body of the request is returned; otherwise, the function throws an exception.
+     */
+    private fun <T> execute(call: Call<T>): T {
```

Comment:
Being this a direct copy of the one from NexusIQ, I'm going to imagine that some Helper, etc... would be a better place for this. Any preference on where?

Reply:
In general the `NexisIq` and `OssIndex` classes have quite some code in common, so maybe it'd make sense to move that common code (incl. this function) to a `SonatypeAdvisor` base class? But the answer also depends on whether we should make use of the *ossindex-public* library instead.